[Feat] 음식점 정보 수정, 음식점 삭제 API 추가

9yuhyeon opened the pull request:
**음식점 수정**
- 음식점 정보 수정 시 음식점 등록할 때 반환되는 ownerAuthToken(점주 인증 토큰)을 필수 입력하여 DB에 저장된 해당 음식점 점주의 인증 토큰과 검증 후 수정 작업 진행
- 음식점 정보 수정은 OWNER, MASTER만 수정할 수 있도록 구현
- 기존 RestaurantRequest에서 음식점 등록, 수정 시 영업시간(시작/종료) 필드는 정규식 검사와 함께 바로 LocalTime 타입으로 변환했으나 형식에 맞지 않는 값 입력 시 지정한 메세지가 반환되지 않고 500, 에러로 반환되어 RestaurantRequest에서 예외 처리가 필요한 상황
- 해당 필드는 String으로 정규식 검사를 마친 후 서비스에 전달하여 서비스에서 LocalTime 타입으로 변환(단, 수정의 경우 해당 값이 NULL일 수 있어 타입 변환 시점을 set 메서드에서 null 체크 후 변환)

**음식점 삭제**
- 음식점 삭제 시 MASTER만 삭제할 수 있도록 구현
- 음식점 삭제는 Hard Delete가 아닌 Soft Delete로 삭제할 수 있도록 구현

P.S 작업 진행 속도가 느려 죄송합니다.. 예상치 못한 에러를 해결하는데 시간을 너무 많이 썼네요ㅠㅠ
주말 동안 최대한 작업 속도를 끌어올려 보도록 하겠습니다!

9yuhyeon committed:
chore(Restaurant): 음식점 정보 수정 API 추가
- 음식점 수정에 필요한 점주 인증 토큰을 음식점 등록 시 반환(관리자가 점주에게 직접 전달 가정)
- 음식점 점주 인증 토큰을 음식점 엔티티에 컬럼 추가
- 컨트롤러에 요청 데이터가 있는 API는 입력값 검증을 위한 @Valid 추가

9yuhyeon committed:
chore(Restaurant): 영업 시간(시작,종료) 필드를 서비스에서 LocalTime으로 변환하도록 수정
- LocalTime의 경우 기존에는 requestDto에서 입력 값 검증 & 타입 변환 하고 있었음
- 잘못된 형식의 값 입력 시 @Pattern에 지정한 message가 반환되지 않고 클라이언트에는 500, 서버 에러 / 서버에는 DateTimeParseException 발생
- requestDto에서는 클라이언트가 보낸 시간 입력값을 String으로 받아 정규표현식으로 검증 후 서비스로 전달
- 서비스에서 형식이 검증된 문자열을 LocalTime으로 변환
- 응답할 때는 DB에서 꺼낸 LocalTime을 String으로 변환 후 반환

9yuhyeon committed:
chore(Restaurant): 음식점 수정 API LocalTime 변환 시점 변경
- 음식점 수정의 경우 점주 인증 토큰 제외한 나머지 필드는 선택적 요소임
- 기존에는 서비스의 클라이언트의 요청 데이터를 수정하는 메서드로 전달하는 시점에서 String -> LocalTime 타입 변환
- 클라이언트가 요청 시 입력한 값이 null일 경우 NullPointerException 발생하여 변환 불가
- String -> LocalTime 변환 시점을 수정 메서드 내부에서 null 체크 후 수정할 수 있도록 수정

9yuhyeon committed:
chore(Restaurant): 미사용 import 제거

9yuhyeon committed:
chore(Restaurant): null 체크 오타 수정

9yuhyeon committed:
chore(Restaurant): ownerAuthToken 기본 값을 공백으로 초기화 제거

JihwanB commented on src/main/java/com/order/orderlink/common/auth/util/TokenGenerator.java:
p5; 개인적인 의견입니다만, 이거는 공통적인 인증/인가가 아닌 음식점 등록시에 발급되는 거니까 pakage를 auth보다 restaurant 쪽으로 가져가는건 어떨까 싶습니다.

9yuhyeon replied:
안그래도 이 부분을 어떤 패키지에 둘 지 고민이 많았는데 restaurant에 옮기도록 하겠습니다!

JihwanB commented on src/main/java/com/order/orderlink/restaurant/application/RestaurantService.java:
```diff
@@ -18,122 +12,171 @@
+                .build();
+    }
+
+    // Convert : Restaurant -> RestaurantDto
```
p5; 이거는 람다 표현식 ( (restaurant) -> {...} ) 사용하면 별도의 변환 메서드 없어도 될 것 같습니다!

9yuhyeon replied:
네 Entity를 DTO로 매핑하는 메서드 제거하고 람다로 DTO 객체 바로 build 하도록 수정했습니다.

JihwanB commented on src/main/java/com/order/orderlink/restaurant/application/RestaurantService.java:
```diff
@@ -18,122 +12,171 @@
+    }
+
+    // Convert : Food -> RestaurantFoodDto
+    private GetRestaurantFoodDto convertToRestaurantFoodDto(Food food) {
```
p5; 이거도 마찬가지로 람다 표현식( (food) -> {...} )으로 대체 가능할 것 같습니다!

JihwanB commented on src/main/java/com/order/orderlink/restaurant/application/dtos/RestaurantRequest.java:
```diff
@@ -24,22 +20,46 @@ public static class Create {
+        private String businessRegNum;
 
+        @NotBlank(message = "토큰은 필수 입력이며, 공백을 포함할 수 없습니다.")
+        @Pattern(regexp = "^[A-Za-z0-9_-]{22}$", message = "음식점 인증을 위한 점주 인증 토큰이 올바르지 않습니다.")
```
p5; 메세지를 좀 더 간결하고 이해하기 쉽게 "점주 인증 토큰이 올바르지 않습니다" 정도로만 해도 좋지 않을까 싶습니다!

9yuhyeon replied:
네 메세지를 조금 더 간결하게 줄이도록 하겠습니다.

JihwanB commented on src/main/java/com/order/orderlink/restaurant/domain/Restaurant.java:
```diff
@@ -97,4 +101,40 @@ public void updateRatingCount(Integer newRatingCount) {
+        }
+
+        if (ownerName != null && !ownerName.trim().isEmpty()) {
+            this.ownerName = ownerName;
```
p5; 다른 String 필드들과 달리 ownerName만 trimming을 해 주신 이유가 있으신가요?

9yuhyeon replied:
- 다른 String 필드(음식점명, 주소, 음식점 소개)는 공백이 들어갈 수 있을 것 같아 trimming을 설정하지 않았습니다.
- phone(연락처), businessRegNum(사업자 등록 번호)은 클라이언트 요청 시 컨트롤러에서 매핑되는 dto 객체 내부에 정규표현식 검증이 이루어집니다.  
- ownerName(점주명)은 사람 이름에 공백을 허용하는 것이 적합하지 않다고 판단하여 trimming을 했습니다.
- openTime과 closeTime은 dto에서 String으로 받으며 dto에서 LocalTime으로 변환 시 사용자가 잘못된 값을 입력했을 경우 지정한 message가 반환되지 않고 code 500으로 반환되어 서비스 레벨에서 변환하도록 설계했습니다.

JihwanB replied:
p5; 넵 무슨 말씀이신지 잘 이해했습니다!
다만 제가 코드상으로 이해한 바로는, 요청으로 들어온 데이터에 실제로 trim()을 해서 반영한다기 보다, trim()된 결과값을 가지고 요청값이 비어있는지를 조건체크 하는것으로 보여집니다.
말씀하신 것처럼 음식점명, 주소, 음식점 소개 같은 필드들도 공백이 포함이 될 수 있습니다만, 이 조건식에서 trim()을 하는것이 실제 데이터에서 공백을 빼고난 후에 저장을 하는것이 아니라,  공백을 제외한 후에 남아있는 데이터가 있는것인지 확인하는것으로 판단됩니다(즉, 요청값이 null은 아니지만 공백만을 가지고 있지는 않은지 확인하는 작업). 요청에서 String 타입에 Validation으로 NotBlank를 사용한다면 문제가 없겠지만, 그게 아닌 필드가 있다면 이 부분을 한번 확인해봐도 좋을 것 같습니다!

JihwanB replied:
> * 다른 String 필드(음식점명, 주소, 음식점 소개)는 공백이 들어갈 수 있을 것 같아 trimming을 설정하지 않았습니다.
> * phone(연락처), businessRegNum(사업자 등록 번호)은 클라이언트 요청 시 컨트롤러에서 매핑되는 dto 객체 내부에 정규표현식 검증이 이루어집니다.
> * ownerName(점주명)은 사람 이름에 공백을 허용하는 것이 적합하지 않다고 판단하여 trimming을 했습니다.
> * openTime과 closeTime은 dto에서 String으로 받으며 dto에서 LocalTime으로 변환 시 사용자가 잘못된 값을 입력했을 경우 지정한 message가 반환되지 않고 code 500으로 반환되어 서비스 레벨에서 변환하도록 설계했습니다.

@9yuhyeon 
p5; 넵 무슨 말씀이신지 잘 이해했습니다!
다만 제가 코드상으로 이해한 바로는, 요청으로 들어온 데이터에 실제로 trim()을 해서 반영한다기 보다, trim()된 결과값을 가지고 요청값이 비어있는지를 조건체크 하는것으로 보여집니다.

말씀하신 것처럼 음식점명, 주소, 음식점 소개 같은 필드들도 공백이 포함이 될 수 있습니다만, 이 조건식에서 trim()을 하는것이 실제 데이터에서 공백을 빼고난 후에 저장을 하는것이 아니라,  공백을 제외한 후에 남아있는 데이터가 있는것인지 확인하는것으로 판단됩니다(즉, 요청값이 null은 아니지만 공백("" 혹은 "  ")만을 가지고 있지는 않은지 확인하는 작업). 요청에서 String 타입에 Validation으로 NotBlank를 사용한다면 문제가 없겠지만, 그게 아닌 필드가 있다면 이 부분을 한번 확인해봐도 좋을 것 같습니다!

즉 지금 같은 경우 업데이트라서 일부 필드에만 값이 들어올 수 있기 때문에 Request.Update 쪽에 NotNull이나 NotBlank 같은 검증이 안걸려 있습니다. 그렇기에 ownerName외에 다른 String 타입 필드들도 trim()을 하여서 비어있지는 않은지 추가적으로 검증하는 것도 좋아보입니다!

9yuhyeon replied:
확인해보니 다른 필드들에 대해서는 빈 값("")과 공백(" ")을 체크하지 않는군요..
디테일한 부분까지 리뷰해 주셔서 감사합니다!
다른 String 필드들도 빈 값과 공백을 검증하도록 수정하겠습니다.

JihwanB reviewed (commented):
수고하셨습니다!!

9yuhyeon committed:
chore(Restaurant): 점주 인증 토큰 관련 메세지 간결하게 수정

9yuhyeon committed:
chore(Restaurant): TokenGenerator 클래스 위치 이동
common.auth.util -> restaurant.application

9yuhyeon committed:
chore(Restaurant): 음식점 전체 조회, 특정 음식점 조회 서비스 로직에서 DTO 객체 매핑 방법 수정
- 기존 : convert method 사용
- 변경 : 람다 사용하여 dto 객체 바로 build

9yuhyeon commented:
@JihwanB 
리뷰 남기신 내용 모두 수정하여 push 했습니다!

JihwanB reviewed (approved):
수고하셨습니다! 제가 위에 규현님 댓글아래에 추가로 작성한 댓글만 한번 확인 부탁드리겠습니다!

JihwanB commented on src/main/java/com/order/orderlink/common/enums/ErrorCode.java:
```diff
@@ -9,6 +9,8 @@ public enum ErrorCode {
 
 	// AUTH
 	TOKEN_INVALID(HttpStatus.UNAUTHORIZED, "토큰이 유효하지 않습니다."),
+	OWNER_AUTH_TOKEN_REQUIRED(HttpStatus.UNAUTHORIZED, "음식점 정보 수정을 위해 인증 토큰은 필수 입력입니다."),
```
p5; 아! 이 부분 메시지도 동일하게 맞춰주시면 더 좋을 것 같습니다! 예를 들어, "점주 인증 토큰은 필수입니다." 이런식으로요!
제가 지금 못찾은 것일수도 있는데, OWNER_AUTH_TOKEN_REQUIRED랑 INVALID_OWNER_AUTH_TOKEN이 사용되는 곳을 못찾겠습니다. 만약 불필요하다면 그냥 빼주셔도 괜찮습니다! 규현님께서 편하신 대로 해주시면 될 것 같습니다!

9yuhyeon replied:
해당 메세지가 사용되는 곳은 ErrorCode와 Request.Update 두 곳이 전부입니다.
말씀 주신 대로 메세지는 간결하게 통일 시키도록 하겠습니다 :)

9yuhyeon committed:
chore(Restaurant): 음식점 정보 수정 시 데이터 형식 검증 수정
- 기존 ownerName만 null, 빈 값(""), 공백(" ")을 체크함.
- 모든 String 필드를 null, 빈 값(""), 공백(" ")을 체크하도록 수정

JihwanB reviewed (approved):
수고하셨습니다!!

9yuhyeon committed:
feat(Restaurant): 음식점 삭제 API 추가
- MASTER 권한만 음식점 삭제할 수 있도록 설계
- Hard Delete가 아닌 Soft Delete로 음식점이 삭제되도록 처리

9yuhyeon committed:
chore(Restaurant): 음식점의 음식리스트 cascade, orphanRemoval 옵션 적용

9yuhyeon committed:
chore(Restaurant): 음식점 삭제 시 softDelete 메서드에 restaurantId 대신 userId 전달하도록 수정
